crates/datasources/src/json/table.rs

```diff
@@ -0,0 +1,135 @@
+    source_url: DatasourceUrl,
+) -> Result<Arc<dyn TableProvider>, JsonError> {
+    let path = source_url.path();
+
+    let store = store_access.create_store()?;
+
+    // assume that the file type is a glob and see if there are
+    // more files...
+    let mut list = store_access.list_globbed(&store, path.as_ref()).await?;
+
+    if list.is_empty() {
+        return Err(JsonError::NotFound(path.into_owned()));
+    }
+
+    // for consistent results, particularly for the sample, always
+    // sort by location
+    list.sort_by(|a, b| a.location.cmp(&b.location));
+
+    let mut data = Vec::new();
+    for obj in list {
+        let blob = store.get(&obj.location).await?.bytes().await?.to_vec();
+        let dejson = serde_json::from_slice::<serde_json::Value>(blob.as_slice())?.to_owned();
+        push_unwind_json_values(&mut data, dejson)?;
+    }
+
+    let mut field_set = indexmap::IndexMap::<String, DataType>::new();
+    for obj in &data {
+        for key in obj.keys() {
+            if field_set.contains_key(key) {
+                continue;
+            }
+            field_set.insert(key.to_string(), type_for_value(obj.get(key).unwrap()));
+        }
+    }
+
+    let schema = Arc::new(Schema::new(
+        field_set
+            .iter()
+            .map(|(k, v)| Field::new(k, v.to_owned(), true))
+            .collect::<Vec<_>>(),
+    ));
+
+    let chunks = data
+        .chunks(100)
+        .map(|chunk| -> Arc<dyn PartitionStream> {
+            Arc::new(JsonPartitionStream::new(
+                schema.clone(),
+                chunk.to_vec().to_owned(),
+            ))
+        })
+        .collect::<Vec<_>>();
+
+    Ok(Arc::new(StreamingTable::try_new(schema.clone(), chunks)?))
+}
+
+fn push_unwind_json_values(
+    data: &mut Vec<Map<String, Value>>,
+    val: Value,
+) -> Result<(), JsonError> {
+    match val {
+        Value::Array(vals) => {
+            for v in vals {
+                match v {
+                    Value::Object(doc) => data.push(doc),
+                    Value::Null => data.push(Map::new()),
+                    _ => {
+                        return Err(JsonError::UnspportedType(
+                            "only objects and arrays of objects are supported",
+                        ))
+                    }
+                }
+            }
+        }
+        Value::Object(doc) => data.push(doc),
+        Value::Null => data.push(Map::new()),
+        _ => {
+            return Err(JsonError::UnspportedType(
+                "only objects and arrays of objects are supported",
+            ))
+        }
+    };
+    Ok(())
+}
+
+fn type_for_value(value: &Value) -> DataType {
+    match value {
+        Value::Array(v) => {
+            if v.is_empty() {
+                DataType::List(Arc::new(Field::new("", DataType::Null, true)))
+            } else {
+                DataType::List(Arc::new(Field::new(
+                    "",
+                    type_for_value(v.first().unwrap()),
```

Comment:
This'll definitely cause some issues later on. I did this in the polars json parser & we ended up getting a bunch of issues from parsing errors due to schema mismatching. We ended up needing to infer the super type for the values in the array. 

for example, an array of `[null, "some string", "some other string"]` will result in `List<Null>` instead of `List<Utf8>`.

similarly, an array of `[0, 1.0, 2.2]` will get parsed as `List<Int64>` instead of `List<Float64>`

Reply:
At a minimum, we should error out if the values in the array are of different types.

Reply:
yeah, I thought of this and decided to not care initially. There are a lot of edge cases that we'll continue to not handle very well (`[nul, "foo", 1]` plus arrays of objects...

I can do something replaces nulls with something that isn't a null (that'd cover a lot of cases) sort of everything else gets much harrier.